max7219/max7219.go

```diff
@@ -0,0 +1,458 @@
+	_REGISTER_SCAN_LIMIT   byte = 0xb
+	_REGISTER_SHUTDOWN     byte = 0xc
+	_REGISTER_DISPLAY_TEST byte = 0xf
+
+	// Value to write to a Code B Font decoded register to blank out the
+	// digit.
+	ClearDigit byte = 0x0f
+	// Value to write for a minus sign symbol
+	MinusSign byte = 0x0a
+	// To turn the decimal point on for a digit display, OR the value of the
+	// digit with DecimalPoint
+	DecimalPoint byte = 0x80
+
+	// DecodeB is used for numeric segment displays. E.G. given a binary 0,
+	// it would turn on the appropriate segments to display the character 0.
+	DecodeB DecodeMode = 0xff
+	// DecodeNone is RAW mode, or not decoded. For each byte, bits that are
+	// one are turned on in the matrix, and bits that are 0 turn off the
+	// led at that row/column.
+	DecodeNone DecodeMode = 0
+)
+
+// Type for a Maxim MAX7219/MAX7221 device.
+type Dev struct {
+	conn spi.Conn
+	mu   sync.Mutex
+	// decode mode for all data registers
+	decode DecodeMode
+	// units is the number of 7219 units daisy-chained together.
+	units int
+	// The number of digits (or the NxN matrix size) in this display.
+	digits byte
+	// charset is the character set for matrix display use. The first index
+	// is the code point (e.g. 0x20=32=space. The second index is the 8 byte
+	// raster values for each line in the matrix.
+	glyphs [][]byte
+}
+
+// emptyBytes creates a slice of empty bytes (digit values or byte values)
+// for the display.
+func (d *Dev) emptyBytes() []byte {
+	b := make([]byte, d.digits)
+	if d.decode == DecodeB {
+		for ix := range b {
+			b[ix] = ClearDigit
+		}
+	}
+	return b
+}
+
+// init puts the display in the default mode. The default is to
+// clear the display, set intensity to middle, and set Decode
+// to DecodeB for a single unit, or DecodeNone for multi-unit.
+func (d *Dev) init() {
+	var initCommands = [][]byte{
+		{_REGISTER_DISPLAY_TEST, 0x0},
+		{_REGISTER_SHUTDOWN, 0x00},
+		{_REGISTER_INTENSITY, 0x08},
+		{_REGISTER_SCAN_LIMIT, d.digits - 1},
+		{_REGISTER_SHUTDOWN, 0x01}}
+
+	d.mu.Lock()
+	for _, cmd := range initCommands {
+		err := d.sendCommand(cmd[0], cmd[1])
+		if err != nil {
+			log.Println(err)
+			break
+		}
+	}
+	d.mu.Unlock()
+	if d.units == 1 {
+		_ = d.SetDecode(DecodeB)
+	} else {
+		_ = d.SetDecode(DecodeNone)
+	}
+
+	_ = d.Clear()
+}
+
+// sendCommand writes to a data register or command register.
+// Data registers are 1-8, and command registers are > 8.
+// If multiple units are daisychained together, the command
+// is repeated and sent to all units.
+func (d *Dev) sendCommand(register, data byte) error {
+	w := make([]byte, d.units*2)
+	for ix := range d.units {
+		w[ix*2] = register
+		w[ix*2+1] = data
+	}
+	return d.conn.Tx(w, nil)
+}
+
+// NewSPI creates a new Max7219 using the specified spi.Port. units is the number
+// of Max7219 chips daisy-chained together. numDigits is the number of digits
+// displayed.
+func NewSPI(p spi.Port, units, numDigits int) (*Dev, error) {
+	if numDigits <= 0 {
```

Comment:
`|| numDigits/units > 8`

but the units <= 0 check would have to be done first.